Refactor exception and middleware handling

daniel-myparcel opened the pull request:
## What Changed
This is a refactor that adds `ExceptionMapper` and `MiddlewareMapper`, both class that can be passed as a `callable` in `withExceptions` and `withMiddleware` respectively in `boostrap/app.php`.

### ExceptionMapper
Overwrites the default Laravel exception handling in a less intrusive way than the old `Handler.php` class which would overwrite the `render` function of the exception handler in Laravel

### MiddlewareMapper
To keep things clean we decided to use this opportunity to also migrate the middleware initialization to this form to align with the current method descibed in Laravel 11.

### README Update
The following error occured when installing the project using the readme.

```shell
➜ docker run -v $(pwd):/app --rm -w /app php:8.2-cli vendor/bin/phpunit
This version of PHPUnit requires PHP >= 8.3.
You are using PHP 8.2.27 (/usr/local/bin/php).
```

This was fixed by updating the `README.md`

This change entirely eliminates the need for a seperate Kernal file for middleware.

![refactor](https://media.tenor.com/eQ8OVVGD5rIAAAAM/refactor.gif)

## Related ticket
https://myparcelcombv.atlassian.net/browse/MP-7141?atlOrigin=eyJpIjoiNWZiOTYzNTFjN2YxNDY0MWIxZGVjYTIyMmQ4MDk0MDMiLCJwIjoiaiJ9

yoan-myparcel reviewed (changes requested):
I suggest we get rid off the handler altogether. The integrations that need the old handler can pin earlier version of _marketplace-commons_.

daniel-myparcel commented:
I see your point. I removed the Exception handling logic and renamed the class to ExceptionMapper which is now only used to expose static methods to map exceptions to responses.

yoan-myparcel commented on src/Exceptions/ExceptionMapper.php:
```diff
@@ -0,0 +1,105 @@
+        return [
+            'Content-Type' => 'application/vnd.api+json',
+        ];
+    }
```
Is there a reason for all these methods to stay static?

yoan-myparcel commented on src/Http/Middleware/Authenticate.php:
Haha looking at this middleware we can get rid of it altogether 🔥 There isn't a situation in marketplaces where we need it, it must have been left in the skeleton since the quickstart app.

daniel-myparcel replied:
Nice! Since we are already cleaning this up, what about the other middleware that existed in the commons folder before? Is that used by some marketplace integrations or can they also be removed?

daniel-myparcel commented on tests/Exceptions/ExceptionMapperTest.php:
```diff
@@ -0,0 +1,86 @@
+        $this->assertEquals(300, $error['status']);
+        $this->assertEquals('Some error', $error['detail']);
+        $this->assertArrayHasKey('trace', $error);
+    }
```
I was trying to write a test that ensure the correct status code is picked depending on the different exceptions it can encounter. However, I could not manage to overwrite the default exception values for `code` and `message`. Since this test doesn't didn't push it but here is what I have so far:
```php
    public function test_it_maps_default_exception_status_correctly(): void
    {
        $exception = new Exception();

        $response = $this->exceptionMapper->getDefaultExceptionStatus($exception);

        $this->assertEquals(500, $response);

        $psrResponseMock = Mockery::mock(ResponseInterface::class);
        $psrResponseMock->allows('getStatusCode')->andReturn(400);
        $psrResponseMock->allows('getBody')->andReturn('Some error');


        $exception = new class(new Response($psrResponseMock)) extends RequestException {
            protected function prepareMessage(Response $response): string
            {
                return 'hello';
            }
        };
        
        $response = $this->exceptionMapper->getDefaultExceptionStatus($exception);

        $this->assertEquals(400, $response);

    }
```

I tried to mock different classes with the same outcome being that `getCode` always returns 500, regardless of the value set to `code` in the constructor.

@yoan-myparcel Let me know if you think we should just drop this test, or if you have any ideas on how to still mock this class.

yoan-myparcel replied:
I made some changes to this section of the tests actually, mainly the access to methods such as _getDefaultExceptionStatus_ - this method is declared public, meanwhile it is being used privately. Therefore, the test has to change to accommodate that. I got a conflict with this actually in my PR. I suggest you take a look at how I did the testing in my PR, and then you can expand with this additional test case because it is still not covered yet, even in my PR.
